### ISSUE TITLE
buildcache push: local buildcache to avoid re-tarballing and preserve shasums 

### ISSUE THREAD

**Issue by haampie** [LLNL/spack#47122]
### Summary

`spack buildcache push` **always** creates new tarballs from install prefixes of specs to be pushed.

This is problematic because it's

1. inefficient when pushing the same spec to multiple buildcaches (for example when autopush is enabled on multiple buildcaches)
2. inefficient when `buildcache push` fails due to some network issue and has to be run again
3. inefficient when redistributing specs using the pattern `spack install` from cache followed by `spack buildcache push`
4. producing bitwise-distinct tarballs when doing `install` from cache followed by `buildcache push` due to relocation:
    1. unnecessary loss of 1-1 mapping between dag hash and shasum of tarball
    2. original install tree padding can get lost due to relocation, which is a footgun
    3. inefficient for stores (oci registries) and file systems (zfs) that do content deduplication

My suggestion is to maintain a local filesystem based buildcache, to which the tarballs are pushed prior to the selected build cache, and keep an index for it in a secure location like `~/.spack` (the mapping shasum -> dag hash has to be trusted).

Similarly, `spack install` from a remote build cache would cache tarballs in this local buildcache.

This way `spack buildcache push` can take a shortcut when it encounters dag hashes already in the local build cache.

Also `spack buildcache sync` could make use of this and be an actually useful command to copy binaries from A to B (should simplify syncing all combinations of [src, dest] x [local fs, non-oci, oci])

TL;DR: this all looks very similar to `docker push` and `docker pull`: store original images (i.e. bunch of tarballs) additionally to just extracting them.

This is related to @kwryankrattiger's issue about making file system based build caches content addressable + restructuring build cache indices: we just need ordinary caching of downloaded binary tarballs by shasum, and then an index.json (probably in a non-standard dir) that registers the spec -> shasum mapping.

**Comment by kwryankrattiger** [comment]
I think this could be pretty helpful for some workflows.

I am a little concerned about how fast a local binary cache could grow on a local FS. There is a `spack clean --cache` that probably helps with this a bit. I think in addition to this having some tools to query how big a cache is and only clearing the local binary cache vs the local source cache, etc. is also a good idea.
